boot/bootstate20.go

```diff
@@ -356,31 +356,103 @@ func (envbks *envRefExtractedKernelBootloaderKernelState) setNextKernel(sn snap.
 }
 
 //
-// kernel snap methods
+// bootStateUpdate for 20 methods
 //
 
-// bootState20Kernel implements the bootState and bootStateUpdate interfaces for
-// kernel snaps on UC20. It is used for setNext() and markSuccessful() - though
-// note that for markSuccessful() a different bootStateUpdate implementation is
-// returned, see bootState20MarkSuccessful
-type bootState20Kernel struct {
-	bks bootloaderKernelState20
+type bootCommitTask func() error
+
+// bootState20Kernel implements the bootStateUpdate interface for both kernel
+// and base snaps on UC20.
+type bootStateUpdate20 struct {
+	// tasks to run before the modeenv has been written
+	preModeenvTasks []bootCommitTask
+
+	// the modeenv that was read from the state
+	modeenv *Modeenv
+
+	// the modeenv that will be written out in commit
+	writeModeenv *Modeenv
+
+	// tasks to run after the modeenv has been written
+	postModeenvTasks []bootCommitTask
+}
+
+func newBootStateUpdate20(m *Modeenv) *bootStateUpdate20 {
+	u20 := &bootStateUpdate20{
+		writeModeenv: &Modeenv{},
+	}
+	// TODO: is this too silly?
```

Comment:
yes, also it won't work if Modeenv grows maps inside,  better to have an explicit copy method

Reply:
#9097 now implements a proper Duplicate method that we use here